python/ray/tune/sync_client.py

```diff
@@ -0,0 +1,238 @@
+
+    def sync_down(self, source, target):
+        """Syncs down from source to target.
+
+        Args:
+            source (str): Source path.
+            target (str): Target path.
+
+        Returns:
+            True if sync initiation successful, False otherwise.
+        """
+        raise NotImplementedError
+
+    def delete(self, target):
+        """Deletes target.
+
+        Args:
+            target (str): Target path.
+
+        Returns:
+            True if delete initiation successful, False otherwise.
+        """
+        raise NotImplementedError
+
+    def wait(self):
+        """Waits for current sync to complete, if asynchronously started."""
+        pass
+
+    def reset(self):
+        """Resets state."""
+        pass
+
+
+class FunctionBasedClient(SyncClient):
+    def __init__(self, sync_up_func, sync_down_func, delete_func=None):
+        self.sync_up_func = sync_up_func
+        self.sync_down_func = sync_down_func
+        self.delete_func = delete_func or noop
+
+    def sync_up(self, source, target):
+        self.sync_up_func(source, target)
+        return True
+
+    def sync_down(self, source, target):
+        self.sync_down_func(source, target)
+        return True
+
+    def delete(self, target):
+        self.delete_func(target)
+        return True
+
+
+NOOP = FunctionBasedClient(noop, noop)
+
+
+class CommandBasedClient(SyncClient):
+    def __init__(self,
+                 sync_up_template,
+                 sync_down_template,
+                 delete_template=noop_template):
+        """Syncs between two directories with the given command.
+        Arguments:
+            sync_up_template (str): A runnable string template; needs to
+                include replacement fields '{source}' and '{target}'.
+            sync_down_template (str): A runnable string template; needs to
+                include replacement fields '{source}' and '{target}'.
+            delete_template (Optional[str]): A runnable string template; needs
+                to include replacement field '{target}'. Noop by default.
+        """
+        self._validate_sync_string(sync_up_template)
+        self._validate_sync_string(sync_down_template)
+        self.sync_up_template = sync_up_template
+        self.sync_down_template = sync_down_template
+        self.delete_template = delete_template
+        self.logfile = None
+        self.cmd_process = None
+
+    def set_logdir(self, logdir):
+        """Sets the directory to log sync execution output in.
+        Args:
+            logdir (str): Log directory.
+        """
+        self.logfile = tempfile.NamedTemporaryFile(
+            prefix="log_sync", dir=logdir, suffix=".log", delete=False)
+
+    def sync_up(self, source, target):
+        return self._execute(self.sync_up_template, source, target)
+
+    def sync_down(self, source, target):
+        return self._execute(self.sync_down_template, source, target)
+
+    def delete(self, target):
```

Comment:
Why does it make sense that the SyncClient also handles deletes?

Reply:
Hmm I don't have a strong reason for why/why not, just seems intuitive to me and is the simplest way to implement. lmk if there's another place to put this functionality that makes more sense.

Reply:
It seems to overload this class with behavior that is not related to syncing. How is the delete function even used? Is this behavior actually used in Tune yet?  If not, we should take it out. Can approve and merge  once so.

Reply:
done. i'll add it to the other PR and we can discuss it there